librz/reg/profile.c

```diff
@@ -1,142 +1,397 @@
-	reg->bits |= item->size;
+	def->packed = parse_size(rz_list_get_n(tokens, 4));
+	if (def->packed == UT32_MAX) {
+		RZ_LOG_WARN("Invalid register packed size.\n");
+		goto reg_parse_error;
+	}
+
+	if (!parse_offset(rz_list_get_n(tokens, 3), def)) {
+		RZ_LOG_WARN("Invalid register offset.\n");
+		goto reg_parse_error;
+	}
 
-	// This is optional
-	if (n == 6) {
-		if (*tok[5] == '#') {
+	// Comments and flags are optional
+	if (rz_list_length(tokens) == 6) {
+		const char *comment_flag = rz_list_get_n(tokens, 5);
+		rz_return_val_if_fail(comment_flag, false);
+		if (comment_flag[0] == '#') {
 			// Remove # from the comment
-			item->comment = strdup(tok[5] + 1);
+			def->comment = strdup(comment_flag + 1);
 		} else {
-			item->flags = strdup(tok[5]);
+			def->flags = strdup(comment_flag);
 		}
 	}
 
-	item->arena = type2;
-	if (!reg->regset[type2].regs) {
-		reg->regset[type2].regs = rz_list_newf((RzListFree)rz_reg_item_free);
-	}
-	rz_list_append(reg->regset[type2].regs, item);
-	if (!reg->regset[type2].ht_regs) {
-		reg->regset[type2].ht_regs = ht_pp_new0();
+	rz_list_append(def_list, def);
+
+	return true;
+
+reg_parse_error:
+	rz_reg_profile_def_free(def);
+	return false;
+}
+
+/**
+ * \brief Parses a register profile string. Each line is either a register alias or a register definiton.
+ *
+ * A register alias string is of the following form:
+ * "=<alias>  <name>\n"
+ *
+ * A register definition string is of the following form:
+ * "(<sub-type>@)main-type  <name>  .<size>  <byte offset>(.<bit offset>)  <packed>  (# <comment> OR <flags>)\n"
+ *
+ * Elements in "()" are optional.
+ * Each "<...>" token is separated by tab or space characters.
+ *
+ * * alias: Register alias (e.g. PC, A1 etc.)
+ * * name: Register name.
+ * * size: Register size in bits.
+ * * main-type: Register type: gpr, fpr, ctr, flg etc.
+ * * sub-type: The second register type (e.g. xmm@fpu : xmm is sub-type of fpu)
+ * * byte offset: Offset into register profile in bytes.
+ * * bit offset: Offset into the byte offset in bits.
+ * * packed: Packed size of the register in bytes.
+ * * comment: A comment about the register.
+ * * Flags this register holds.
+ *
+ * \param reg Register struct which holds all register items.
+ * \param profile Register profile string.
+ * \return false On failure.
+ * \return true On success.
+ */
+static bool parse_reg_profile_str(RZ_OUT RzList *alias_list, RZ_OUT RzList *def_list, const char *profile_str) {
+	rz_return_val_if_fail(alias_list && def_list && profile_str, false);
+
+	RzList *def_lines = rz_str_split_duplist_n(profile_str, "\n", 0, true);
+	rz_return_val_if_fail(def_lines, false);
+
+	st32 l = 0; // Line number
+	const char *line;
+	bool is_alias = false;
+	RzListIter *it;
+	RzList *toks = NULL;
+	rz_list_foreach (def_lines, it, line) {
+		++l;
+		if (strcmp(line, "") == 0) {
+			continue;
+		}
+		toks = rz_str_split_duplist_n_regex(line, "[[:blank:]]+", 0, true);
+		if (!toks || rz_list_length(toks) == 0) {
```

Comment:
no need to check the null here, since `rz_list_length` accepts it and returns 0, but i guess is better to use directly `rz_list_is_empty`